src/python/pants/backend/rust/lint/rustfmt/rules_integration_test.py

```diff
@@ -0,0 +1,149 @@
+
+from textwrap import dedent
+
+import pytest
+
+from pants.backend.rust.lint.rustfmt.rules import RustfmtFieldSet, RustfmtRequest
+from pants.backend.rust.lint.rustfmt.rules import rules as rustfmt_rules
+from pants.backend.rust.target_types import RustPackageTarget
+from pants.backend.rust.util_rules import toolchains
+from pants.core.goals.fmt import FmtResult
+from pants.core.util_rules import source_files, system_binaries
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Address
+from pants.engine.target import Target
+from pants.testutil.rule_runner import QueryRule, RuleRunner
+
+
+@pytest.fixture()
+def rule_runner() -> RuleRunner:
+    rule_runner = RuleRunner(
+        target_types=[RustPackageTarget],
+        rules=[
+            *rustfmt_rules(),
+            *toolchains.rules(),
+            *source_files.rules(),
+            *system_binaries.rules(),
+            QueryRule(FmtResult, (RustfmtRequest.Batch,)),
+            QueryRule(SourceFiles, (SourceFilesRequest,)),
+        ],
+    )
+    rule_runner.set_options([], env_inherit={"PATH"})
+    return rule_runner
+
+
+GOOD_FILE = dedent(
+    """\
+    use Foo::Bar;
+    fn grok(s: &str) {}
+    """
+)
+
+BAD_FILE = dedent(
+    """\
+    use  Foo::Bar;
+    fn grok(s: &str) {
+    println!("Hello World {}!", s);
+    }
+    """
+)
+
+FIXED_BAD_FILE = dedent(
+    """\
+    use Foo::Bar;
+    fn grok(s: &str) {
+        println!("Hello World {}!", s);
+    }
+    """
+)
+
+
+def run_rustfmt(
+    rule_runner: RuleRunner,
+    targets: list[Target],
+    *,
+    extra_args: list[str] | None = None,
+) -> FmtResult:
+    rule_runner.set_options(extra_args or (), env_inherit={"PATH"})
+    field_sets = [RustfmtFieldSet.create(tgt) for tgt in targets]
+    input_sources = rule_runner.request(
+        SourceFiles,
+        [
+            SourceFilesRequest(field_set.sources for field_set in field_sets),
+        ],
+    )
+    fmt_result = rule_runner.request(
+        FmtResult,
+        [
+            RustfmtRequest.Batch(
+                "",
+                input_sources.snapshot.files,
+                partition_metadata=None,
+                snapshot=input_sources.snapshot,
+            ),
+        ],
+    )
+    return fmt_result
+
+
+def test_passing(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {"src/lib.rs": GOOD_FILE, "Cargo.toml": "", "BUILD": "rust_package(name='package')"}
+    )
+    tgt = rule_runner.get_target(Address("", target_name="package"))
+    fmt_result = run_rustfmt(rule_runner, [tgt])
+
+    assert fmt_result.stdout.strip() == ""
+    assert fmt_result.output == rule_runner.make_snapshot({"src/lib.rs": GOOD_FILE})
+    assert fmt_result.did_change is False
+
+
+def test_failing(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {"src/lib.rs": BAD_FILE, "Cargo.toml": "", "BUILD": "rust_package(name='package')"}
```

Comment:
Technically you don't need the cargo toml in the dicts anymore

Reply:
I do not see it being an issue having this file in the working dir. A file that would be expected to be there somewhat anyways

Reply:
Not an issue per-se but misleading to the reader IMO. 

Usually we only put what's needed in the sandbox, because it's needed. Otherwise when people read and review the code, they are left to wonder what's important and what isn't. I struggle with that often because I have a small mental stack.

Not a blocker, but is a request. Up to you

Reply:
I see this issue as something to be resolved as development of this backend progresses especially since compilation rules will be next and there will be much to learn about what Rust needs for compilation (or does not need!) then.

Reply:
For that reason, resolving this issue.